win32ss/printing/base/winspool/printers.c

```diff
@@ -384,10 +384,100 @@ EndPagePrinter(HANDLE hPrinter)
 }
 
 BOOL WINAPI
-EnumPrintersA(DWORD Flags, PSTR Name, DWORD Level, PBYTE pPrinterEnum, DWORD cbBuf, PDWORD pcbNeeded, PDWORD pcReturned)
+EnumPrintersA(DWORD Flags, PSTR pName, DWORD Level, PBYTE pPrinterEnum, DWORD cbBuf, PDWORD pcbNeeded, PDWORD pcReturned)
 {
-    TRACE("EnumPrintersA(%lu, %s, %lu, %p, %lu, %p, %p)\n", Flags, Name, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
-    return FALSE;
+    BOOL bReturnValue = FALSE;
+    DWORD cch;
+    PWSTR pwszName = NULL;
+    PSTR pszPrinterName = NULL;
+    PSTR pszServerName = NULL;
+    DWORD i;
+    PPRINTER_INFO_4W ppi4w = NULL;
+    PPRINTER_INFO_4A ppi4a = NULL;
+
+    TRACE("EnumPrintersA(%lu, %s, %lu, %p, %lu, %p, %p)\n", Flags, pName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
+
+    if (pName)
+    {
+        // Convert pName to a Unicode string pwszName.
+        cch = strlen(pName);
+
+        pwszName = HeapAlloc(hProcessHeap, 0, (cch + 1) * sizeof(WCHAR));
+        if (!pwszName)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            ERR("HeapAlloc failed!\n");
+            goto Cleanup;
+        }
+
+        MultiByteToWideChar(CP_ACP, 0, pName, -1, pwszName, cch + 1);
+    }
+
+    bReturnValue = EnumPrintersW(Flags, pwszName, Level, pPrinterEnum, cbBuf, pcbNeeded, pcReturned);
+
+    TRACE("*pcReturned is '%d' and bReturnValue is '%d' and GetLastError is '%ld'.\n", *pcReturned, bReturnValue, GetLastError());
+
+    /* We are mapping two different pointers to the pPrinterEnum pointer here so that */
+    /* we can do in-place conversion. We read the Unicode response from the EnumPrintersW and */
+    /* then we write back the ASCII conversion into the same buffer for our EnumPrintersA output */
+
+    /* mapping to pPrinterEnum for Unicode (w) characters */
+    ppi4w = (PPRINTER_INFO_4W)pPrinterEnum;
```

Comment:
These structures are only used if `Level` is 4. Levels 1, 2 and 5 use different structures and therefore need to be converted differently.

Reply:
Fixed.